File: modules/ai/src/main/java/com/bytedesk/ai/robot/openai_client_example.js
/**
 * Bytedesk OpenAI 兼容接口 Node.js 示例
 * 
 * 安装依赖:
 * npm install openai
 * 
 * 使用方法:
 * node openai_client_example.js
 */

import OpenAI from 'openai';

// 配置 Bytedesk 服务
const BYTEDESK_BASE_URL = 'http://localhost:9003/api/ai/chat/v1';
const BYTEDESK_API_KEY = 'your-api-key'; // 可以是任意值

class BytedeskOpenAIClient {
  constructor(baseURL = BYTEDESK_BASE_URL, apiKey = BYTEDESK_API_KEY) {
    this.openai = new OpenAI({
      apiKey: apiKey,
      baseURL: baseURL
    });
  }

  /**
   * 基本聊天请求
   */
  async chatCompletion(messages, options = {}) {
    try {
      const completion = await this.openai.chat.completions.create({
        model: 'bytedesk-ai',
        messages: messages,
        ...options
      });
      return completion;
    } catch (error) {
      console.error('❌ 聊天请求失败:', error.message);
      return null;
    }
  }

  /**
   * 流式聊天请求
   */
  async streamChatCompletion(messages, options = {}) {
    try {
      const stream = await this.openai.chat.completions.create({
        model: 'bytedesk-ai',
        messages: messages,
        stream: true,
        ...options
      });
      return stream;
    } catch (error) {
      console.error('❌ 流式聊天请求失败:', error.message);
      return null;
    }
  }
}

/**
 * 测试基本聊天功能
 */
async function testBasicChat() {
  console.log('🔄 测试基本聊天功能...');
  
  const client = new BytedeskOpenAIClient();
  
  const messages = [
    { role: 'system', content: '你是一个有帮助的AI助手。' },
    { role: 'user', content: '你好，请介绍一下自己。' }
  ];

  const completion = await client.chatCompletion(messages, {
    temperature: 0.7,
    max_tokens: 200
  });

  if (completion) {
    console.log('✅ 请求成功!');
    console.log(`📝 回复: ${completion.choices[0].message.content}`);
    console.log(`📊 Token 使用:`, completion.usage);
    console.log(`🆔 请求 ID: ${completion.id}`);
  } else {
    console.log('❌ 请求失败');
  }
}

/**
 * 测试多轮对话
 */
async function testMultiTurnConversation() {
  console.log('\n🔄 测试多轮对话...');
  
  const client = new BytedeskOpenAIClient();
  
  const messages = [
    { role: 'system', content: '你是一个记忆力很好的助手。' },
    { role: 'user', content: '我的名字是李四。' },
    { role: 'assistant', content: '你好李四！很高兴认识你。' },
    { role: 'user', content: '我的名字是什么？' }
  ];

  const completion = await client.chatCompletion(messages);

  if (completion) {
    console.log('✅ 多轮对话测试成功!');
    console.log(`📝 回复: ${completion.choices[0].message.content}`);
  } else {
    console.log('❌ 多轮对话测试失败');
  }
}

/**
 * 测试流式聊天
 */
async function testStreamingChat() {
  console.log('\n🔄 测试流式聊天...');
  
  const client = new BytedeskOpenAIClient();
  
  const messages = [
    { role: 'user', content: '请给我讲一个短故事。' }
  ];

  const stream = await client.streamChatCompletion(messages, {
    temperature: 0.8
  });

  if (stream) {
    console.log('✅ 流式请求启动成功!');
    console.log('📝 流式回复: ');
    
    try {
      for await (const chunk of stream) {
        const content = chunk.choices[0]?.delta?.content || '';
        if (content) {
          process.stdout.write(content);
        }
      }
      console.log('\n🏁 流式回复完成');
    } catch (error) {
      console.error(`\n❌ 流式处理错误: ${error.message}`);
    }
  } else {
    console.log('❌ 流式请求失败');
  }
}

/**
 * 测试不同的参数配置
 */
async function testDifferentParameters() {
  console.log('\n🔄 测试不同参数配置...');
  
  const client = new BytedeskOpenAIClient();
  
  const messages = [
    { role: 'user', content: '用一个词描述春天。' }
  ];

  console.log('🌡️  高创造性 (temperature=1.5):');
  const highTemp = await client.chatCompletion(messages, {
    temperature: 1.5,
    max_tokens: 50
  });
  if (highTemp) {
    console.log(`   回复: ${highTemp.choices[0].message.content}`);
  }

  console.log('🧊 低创造性 (temperature=0.1):');
  const lowTemp = await client.chatCompletion(messages, {
    temperature: 0.1,
    max_tokens: 50
  });
  if (lowTemp) {
    console.log(`   回复: ${lowTemp.choices[0].message.content}`);
  }
}

/**
 * 测试错误处理
 */
async function testErrorHandling() {
  console.log('\n🔄 测试错误处理...');
  
  const client = new BytedeskOpenAIClient();
  
  // 测试空消息
  const result = await client.chatCompletion([]);
  
  if (result === null) {
    console.log('✅ 空消息错误处理正确');
  } else {
    console.log('⚠️  空消息未正确处理');
  }
}

/**
 * 检查服务连通性
 */
async function checkServiceConnectivity() {
  console.log('🔍 检查服务连通性...');
  
  try {
    const response = await fetch('http://localhost:9003/actuator/health');
    if (response.ok) {
      console.log('✅ Bytedesk 服务连接正常');
      return true;
    } else {
      console.log('⚠️  Bytedesk 服务状态异常');
      return false;
    }
  } catch (error) {
    console.log(`❌ 无法连接到 Bytedesk 服务: ${error.message}`);
    console.log('请确保服务运行在 localhost:9003');
    return false;
  }
}

/**
 * 主函数
 */
async function main() {
  console.log('🚀 Bytedesk OpenAI 兼容接口测试 (Node.js)');
  console.log('='.repeat(50));

  // 检查服务连通性
  const isConnected = await checkServiceConnectivity();
  if (!isConnected) {
    return;
  }

  // 运行测试
  await testBasicChat();
  await testMultiTurnConversation();
  await testStreamingChat();
  await testDifferentParameters();
  await testErrorHandling();

  console.log('\n' + '='.repeat(50));
  console.log('🎉 所有测试完成!');
  console.log('\n💡 集成提示:');
  console.log('1. 只需要修改 baseURL 就可以使用现有的 OpenAI 代码');
  console.log('2. API Key 可以是任意值，Bytedesk 当前不做验证');
  console.log('3. 支持所有标准的 OpenAI Chat Completions 参数');
  console.log('4. 支持流式和非流式响应');
  console.log('5. 完全兼容 OpenAI JavaScript/TypeScript SDK');
}

// 运行主函数
main().catch(console.error);

File: modules/ai/src/main/java/com/bytedesk/ai/robot/openai_client_example.py
#!/usr/bin/env python3
"""
Bytedesk OpenAI 兼容接口示例
使用标准的 OpenAI Python 客户端库调用 Bytedesk AI 服务

安装依赖:
pip install openai

使用方法:
python openai_client_example.py
"""

import openai
import asyncio
from typing import List, Dict

# 配置 Bytedesk 服务
BYTEDESK_BASE_URL = "http://localhost:9003/api/ai/chat/v1"
BYTEDESK_API_KEY = "your-api-key"  # 可以是任意值，Bytedesk 暂不验证


class BytedeskOpenAIClient:
    """Bytedesk OpenAI 兼容客户端"""
    
    def __init__(self, base_url: str = BYTEDESK_BASE_URL, api_key: str = BYTEDESK_API_KEY):
        self.client = openai.OpenAI(
            api_key=api_key,
            base_url=base_url
        )
    
    def chat_completion(self, messages: List[Dict[str, str]], **kwargs):
        """发送聊天请求"""
        try:
            response = self.client.chat.completions.create(
                model="bytedesk-ai",
                messages=messages,
                **kwargs
            )
            return response
        except Exception as e:
            print(f"❌ 请求失败: {e}")
            return None
    
    def stream_chat_completion(self, messages: List[Dict[str, str]], **kwargs):
        """流式聊天请求"""
        try:
            stream = self.client.chat.completions.create(
                model="bytedesk-ai",
                messages=messages,
                stream=True,
                **kwargs
            )
            return stream
        except Exception as e:
            print(f"❌ 流式请求失败: {e}")
            return None


def test_basic_chat():
    """测试基本聊天功能"""
    print("🔄 测试基本聊天功能...")
    
    client = BytedeskOpenAIClient()
    
    messages = [
        {"role": "system", "content": "你是一个有帮助的AI助手。"},
        {"role": "user", "content": "你好，请介绍一下自己。"}
    ]
    
    response = client.chat_completion(messages, temperature=0.7, max_tokens=200)
    
    if response:
        print("✅ 请求成功!")
        print(f"📝 回复: {response.choices[0].message.content}")
        print(f"📊 Token 使用: {response.usage}")
        print(f"🆔 请求 ID: {response.id}")
    else:
        print("❌ 请求失败")


def test_multi_turn_conversation():
    """测试多轮对话"""
    print("\n🔄 测试多轮对话...")
    
    client = BytedeskOpenAIClient()
    
    messages = [
        {"role": "system", "content": "你是一个记忆力很好的助手。"},
        {"role": "user", "content": "我的名字是张三。"},
        {"role": "assistant", "content": "你好张三！很高兴认识你。"},
        {"role": "user", "content": "我的名字是什么？"}
    ]
    
    response = client.chat_completion(messages)
    
    if response:
        print("✅ 多轮对话测试成功!")
        print(f"📝 回复: {response.choices[0].message.content}")
    else:
        print("❌ 多轮对话测试失败")


def test_streaming_chat():
    """测试流式聊天"""
    print("\n🔄 测试流式聊天...")
    
    client = BytedeskOpenAIClient()
    
    messages = [
        {"role": "user", "content": "请给我讲一个短故事。"}
    ]
    
    stream = client.stream_chat_completion(messages, temperature=0.8)
    
    if stream:
        print("✅ 流式请求启动成功!")
        print("📝 流式回复: ", end="", flush=True)
        
        try:
            for chunk in stream:
                if chunk.choices[0].delta.content is not None:
                    print(chunk.choices[0].delta.content, end="", flush=True)
            print("\n🏁 流式回复完成")
        except Exception as e:
            print(f"\n❌ 流式处理错误: {e}")
    else:
        print("❌ 流式请求失败")


def test_different_parameters():
    """测试不同的参数配置"""
    print("\n🔄 测试不同参数配置...")
    
    client = BytedeskOpenAIClient()
    
    # 测试高创造性参数
    messages = [
        {"role": "user", "content": "用一个词描述春天。"}
    ]
    
    print("🌡️  高创造性 (temperature=1.5):")
    response = client.chat_completion(messages, temperature=1.5, max_tokens=50)
    if response:
        print(f"   回复: {response.choices[0].message.content}")
    
    print("🧊 低创造性 (temperature=0.1):")
    response = client.chat_completion(messages, temperature=0.1, max_tokens=50)
    if response:
        print(f"   回复: {response.choices[0].message.content}")


def test_error_handling():
    """测试错误处理"""
    print("\n🔄 测试错误处理...")
    
    # 测试空消息
    client = BytedeskOpenAIClient()
    response = client.chat_completion([])
    
    if response is None:
        print("✅ 空消息错误处理正确")
    else:
        print("⚠️  空消息未正确处理")


async def test_async_chat():
    """测试异步聊天 (如果需要)"""
    print("\n🔄 测试异步功能...")
    print("ℹ️  当前使用同步客户端，异步功能需要 AsyncOpenAI")


def main():
    """主测试函数"""
    print("🚀 Bytedesk OpenAI 兼容接口测试")
    print("=" * 50)
    
    # 检查服务连通性
    print("🔍 检查服务连通性...")
    try:
        import requests
        response = requests.get("http://localhost:9003/actuator/health", timeout=5)
        if response.status_code == 200:
            print("✅ Bytedesk 服务连接正常")
        else:
            print("⚠️  Bytedesk 服务状态异常")
    except Exception as e:
        print(f"❌ 无法连接到 Bytedesk 服务: {e}")
        print("请确保服务运行在 localhost:9003")
        return
    
    # 运行测试
    test_basic_chat()
    test_multi_turn_conversation()
    test_streaming_chat()
    test_different_parameters()
    test_error_handling()
    
    print("\n" + "=" * 50)
    print("🎉 所有测试完成!")
    print("\n💡 集成提示:")
    print("1. 只需要修改 base_url 就可以使用现有的 OpenAI 代码")
    print("2. API Key 可以是任意值，Bytedesk 当前不做验证")
    print("3. 支持所有标准的 OpenAI Chat Completions 参数")
    print("4. 支持流式和非流式响应")


if __name__ == "__main__":
    main()
